config/config.go

```diff
@@ -19,21 +20,96 @@ type Config struct {
 	StoreURL string   `yaml:"store-url"`
 	Token    string   `yaml:"token"`
 	Launcher Launcher `yaml:"launcher"`
+	filePath string   `yaml:"-"`
 }
 
-// ReadConfig returns parsed config
-func ReadConfig(configPath string) (Config, error) {
-	buf, err := ioutil.ReadFile(configPath)
+func create(configPath string) error {
+	_, err := os.Stat(configPath)
+	// if file exists return nil
+	if err == nil {
+		return nil
+	}
+
+	dir := filepath.Dir(configPath)
+	err = os.MkdirAll(dir, 0777)
+	if err != nil {
+		return err
+	}
+	file, err := os.Create(configPath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	err = yaml.NewEncoder(file).Encode(Config{
+		Launcher: Launcher{
+			Version: "stable",
+			Image:   "screwdrivercd/launcher",
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// New returns parsed config
+func New(configPath string) (Config, error) {
+	err := create(configPath)
+	if err != nil {
+		return Config{}, err
+	}
+
+	file, err := os.Open(configPath)
 	if err != nil {
 		return Config{}, fmt.Errorf("failed to read config file: %v", err)
 	}
 
-	config := Config{}
+	var config = Config{
+		filePath: configPath,
+	}
 
-	err = yaml.Unmarshal(buf, &config)
+	err = yaml.NewDecoder(file).Decode(&config)
 	if err != nil {
-		return Config{}, fmt.Errorf("failed to parse config file: %v\ncontents:\n\n%v", err, string(buf))
+		return Config{}, fmt.Errorf("failed to parse config file: %v", err)
 	}
 
 	return config, nil
 }
+
+// Set preserve sd-local config with new value.
+func (c *Config) Set(key, value string) error {
+	switch key {
+	case "api-url":
+		c.APIURL = value
+	case "store-url":
+		c.StoreURL = value
+	case "token":
+		c.Token = value
+	case "launcher-version":
+		if value == "" {
+			value = "stable"
+		}
+		c.Launcher.Version = value
+	case "launcher-image":
+		if value == "" {
+			value = "screwdrivercd/launcher"
+		}
+		c.Launcher.Image = value
+	default:
+		return fmt.Errorf("invalid key %s", key)
+	}
+
+	file, err := os.OpenFile(c.filePath, os.O_RDWR|os.O_TRUNC, 0666)
```

Comment:
Is O_TRUNC needed? if you want to create a new file, we can use os.O_CREATE.

Reply:
os.O_CREATE is not used because I want to open the conf file if it already exists.
If A is not added, garbage may remain at the end of the file because writing starts from the beginning of the file.
```
# e.g.
# file contents
aaaaaa

# Writes the string of "bbb"
# File after being written is
bbb
aaa
```

Reply:
The reason why we add O_TRUNC is now we write every config file at once. If there are no O_TRUNC, there are possibility that pre settings are left in new setting file and it break the yaml file.
The reason why do not use os.O_CREATE is that the file is already created by this function https://github.com/screwdriver-cd/sd-local/pull/9/files#diff-3baf47c64847a8fb8aaa8cc2e088513bR26 .  

However, actually we can use `os.Create` function. Not so different to call this function and `os.Create`.

If it is difficult to understand we can fix.

Reply:
🙆‍♂